### PR TITLE
Initialize repository

### PR DESCRIPTION
**Pull request by prog893**
Initial code and README to start with.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  function_name     = "${aws_lambda_function.function.arn}"
+  starting_position = "${var.starting_position}"
+}
+
+resource "aws_iam_role_policy_attachment" "xray_access" {
+  policy_arn = "arn:aws:iam::aws:policy/AWSXrayWriteOnlyAccess"
+  role       = "${module.iam.name}"
+}
+
+module "iam" {
```
Comment (toVersus) [inline main.tf]: Be sure to ship this module!

File: variables.tf [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  default     = "time"
+}
+
+variable "tracing_mode" {
+  description = "X-Ray tracing mode (see: https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html )"
+  default     = "PassThrough"
+}
+
+variable "tags" {
+  description = "Resource tags"
```
Comment (toVersus) [inline variables.tf]:
```suggestion
  description = "Tags to Lambda Function"
```

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                "kinesis:ListStreams"
+            ],
+            "Resource": [
+                "${var.kinesis_stream_arn}"
+            ]
+        },
+        {
+            "Effect": "Allow",
+            "Action": [
+                "logs:CreateLogGroup",
```
Comment (toVersus) [inline main.tf]: Why do we need `logs:CreateLogGroup` and `logs:CreateLogStream`? Can we just create these resources in Terraform and restrict Resource access?

Comment (prog893) [reply main.tf]: We can create a LogGroup inside this module, but LogStreams are created dynamically... with unpredictable names.

File: variables.tf (hunk 2 above)
Comment (prog893) [reply variables.tf]: to -> for?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,81 @@
-# aws-lambda_kinesis_to_s3
-Terraform module and Lambda for Kinesis to S3 logs transfer
+# Amazon Kinesis to S3 log transfer Terraform module
```
[the remark this answers is not in the document]
Comment (prog893) [reply README.md]: I guess a version badge or something will do the work?

Should we add a section with supported/tested terraform versions?

File: lambda/main.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+# global S3 client instance
+s3 = boto3.client('s3')
+
+# configure with env vars
+PATH_PREFIX = os.environ['LOG_S3_PREFIX']
+BUCKET_NAME = os.environ['LOG_S3_BUCKET']
+
+# consts
+os.environ['TZ'] = "Asia/Tokyo"
```
[the remark this answers is not in the document]
Comment (prog893) [reply lambda/main.py]: Indeed. So the default value should be UTC then.

File: main.tf (hunk 1 above)
Comment (prog893) [reply main.tf]: 🙈

File: lambda/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (toVersus) [reply lambda/main.py]: Yeah, it looks reasonable.

File: variables.tf (hunk 2 above)
Comment (toVersus) [reply variables.tf]: Oh, thanks! I changed my suggestion.

File: main.tf (hunk 3 above)
Comment (toVersus) [reply main.tf]: I see. So, we can just remove `logs:CreateLogGroup` action from allowed policy.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (toVersus) [reply README.md]: The former idea is my taste! It seems that we need to create a custom badge on [Shields.io](https://shields.io/) like this.

https://img.shields.io/badge/terraform-v0.11.x-brightgreen.svg

[the remark this answers is not in the document]
Comment (prog893) [reply README.md]: Let's go with that badge.

File: main.tf (hunk 1 above)
Comment (prog893) [reply main.tf]: Shipped!

**Comment by prog893** [comment]
@toVersus I have addressed comments above, please re-review! 😸

**Comment by prog893** [comment]
> Modify Usage and Pinning section on README and add sample code (in separate PR?)

Terraform Registry URIs are predictable, so there is no need to update URI in separate PR.

Besides that, procedure above looks good! (P.S. place for internal documentation is still TBD)

**Comment by prog893** [comment]
For initial releases, we could tag v0.0.1 on PR branch before merging to master for testing purposes, and to check how it would look like in Registry. Which is what I am going to do now for real-world tests.

**Comment by toVersus** [comment]
Thanks for sharing your thought, it sounds good! Summarizing the release process as a reference until internal documentation is ready:

1. Review first implementation including sample code pointing to the Terraform Registry URI
1. Put v0.0.1 tag on **PR branch**
1. Register module with Terraform Registry
1. Check the appearance of modules in Terraform Registry
1. Additional E2E test or something using codes tagged with v0.1.0 if needed
1. Get approval and merge it into master
1. Put v1.0.0 tag on **master branch**

Is there something off about these steps?

**Comment by prog893** [comment]
Perfect!

**Comment by prog893** [comment]
I had to do this: 785f427 and d4919e8

So, when we execute curl from within data.external with `-o` option and a relative path, package is downloaded to wherever root module resides, and not cloned module path. Every module instance is copied in its own subfolder in `.terraform/modules`. Here is an example:

```
$ l .terraform/modules
total 4
drwxr-xr-x   8 user  group   256B  7 10 21:17 .
drwxr-xr-x   5 user  group   160B  7 10 21:17 ..
drwxr-xr-x  12 user  group   384B  7 10 21:17 a904f464f450b3deded89377b77bf25c
-rw-r--r--   1 user  group   1.3K  7 10 21:17 modules.json

# add one more module instance

$ terraform plan

Error: Error loading modules: module s3_transfer_second: not found, may need to run 'terraform init'

$ terraform init

Initializing modules...
- module.provider
- module.s3_transfer
- module.s3_transfer_second
  Getting source "github.com/baikonur-oss/terraform-aws-lambda-kinesis-to-s3?ref=init"
- module.s3_transfer.iam
- module.s3_transfer_second.iam

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.18"
* provider.external: version = "~> 1.2"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

$ l .terraform/modules
total 5
drwxr-xr-x   9 user  group   288B  7 10 21:25 .
drwxr-xr-x   5 user  group   160B  7 10 21:17 ..
drwxr-xr-x  11 user  group   352B  7 10 21:25 749080ef9bfd0ede2ac533806b89477a
drwxr-xr-x  12 user  group   384B  7 10 21:17 a904f464f450b3deded89377b77bf25c
-rw-r--r--   1 user  group   1.6K  7 10 21:25 modules.json
```

In order to not cause file overwrite and other nasty stuff, we should download files with static names to separate paths for every module instance. Using dynamic names could help, but would involve generating random strings etc. So, I chose to use `path.module` as prefix for download path, and then this happened.

As commented in d4919e8, `depends_on` with `data.external.package` works for making Terraform wait for download to complete before trying to touch aws_lambda_function resource. However, it does not postpone `base64sha256` execution, which is used for source_code_hash computation (with static or random download filenames, there is no way to detect binary package except hash check). I had to explicitly use `data.external.package.result` *inside* path interpolation in order to enforce wait for download completion. Very dirty hack.

And by the way, external data resource *has to* return some valid JSON, so I am outputting an empty one.

**Comment by prog893** [comment]
@toVersus E2E passed, willing to merge and ship, but... look at the commend below and kindly review.

**Comment by toVersus** [comment]
I think https://github.com/baikonur-oss/terraform-aws-lambda-kinesis-to-s3/commit/785f4279ec651718834f124098fb8fa3d008eeb4 is a well addressed solution for the race condition in case of using multiple Lambda modules at the same time.
However, https://github.com/baikonur-oss/terraform-aws-lambda-kinesis-to-s3/commit/d4919e830dc1dbd415954ade467a2502258f7676 is a really dirty but awesome workaround for dealing with the lack of feature in Terraform. I think about the possibility to fetch pre-uploaded sha256 hash of archived artifact from the GitHub release page but it doesn't address the scheduling problem. (We expect Terraform should wait for the completion of downloading artifact before creating Lambda function resource.)
Let's get back on track. I guess this hack is **not a user-facing part** of this module, so we can ship it as is and make the hack better after initial release. That's probably right, but I predict this hack becomes targets of frequently asked question because this is a core strategy for deploying Lambda function in baikonur-oss project. For the developer, we should write up entire concept of deployment in the internal document sometime while comparing to the existing solution (archiving code in Terraform by using [archive_file](https://www.terraform.io/docs/providers/archive/d/archive_file.html)).

**Comment by prog893** [comment]
E2E passed, merging and shipping v1.0.0.

TODO: write internal docs on release process and lambda module specifics

**Pull request merged**